developer_guides/firmware/llext_modules.rst

```diff
@@ -0,0 +1,78 @@
+.. _llext_modules:
+
+LLEXT Modules
+#############
+
+|SOF| support for loadable modules, using Zephyr LLEXT API.
+
+Zephyr LLEXT API
+****************
+
+Please refer to https://docs.zephyrproject.org/latest/services/llext/index.html
+for detailed documentation. In short, the Zephyr Linkable Loadable Extensions
+(LLEXT) API implements support for run-time loading and unloading of ELF-format
+executable code and data.
+
+SOF Use of the LLEXT API
+************************
+
+SOF has multiple ways to implement loadable modules. LLEXT is one of them.
+With it modules are built as shared or relocatable ELF objects with an addition
+of a cryptographic signature, using any user-supplied key, and a manifest. When
+loaded and instantiated, Zephyr LLEXT functionality is used to dynamically
+resolve module internal as well as SOF external code and data references. In the
+future support for inter-module linking will be added.
+
+Accessing the base firmware from LLEXT modules
+**********************************************
+
+LLEXT modules can access all code and data from the base firmware exported,
+using the ``EXPORT_SYMBOL()`` macro. Therefore writing LLEXT modules isn't very
+different from built-in ones.
+
+Implementing LLEXT modules
+**************************
+
+At the moment only modules, implementing the Module Adapter API
+:ref:`apps-comp-world` are supported.
+
+As explained above, LLEXT modules in general look very similar to native SOF
+code, with the only restriction of having no access to not-exported symbols. If
+any such symbols are needed to LLEXT modules, additionally exporting them can be
+considered.
+
+LLEXT modules should also contain a ``.buildinfo`` section, containing a
+``struct sof_module_api_build_info`` object and a ``.module`` section,
+containing a ``struct sof_man_module_manifest`` object. The latter should also
+contain a pointer to a module entry point function, returning a pointer to the
+module's ``struct module_interface`` instance. See
+src/samples/audio/smart_amp_test_ipc4.c for an example.
+
+A TOML configuration file is needed for building of LLEXT modules too. It is
+generated by the C preprocessor at build time from the same components, as would
+be used for a monolithic build. A small additional file to bring components
+together is needed too, similar to
+src/samples/audio/smart_amp_test_llext/llext.toml.h.
+
+Finally an additional CMakeLists.txt is needed similar to
+src/samples/audio/smart_amp_test_llext/CMakeLists.txt. Currently it contains a
+lot of open-coded instructions, later it will be ported to use Zephyr's
+``add_llext_target()`` function.
+
+With that in place, it is also possible to switch between monolithic and modular
+builds by specifying the module as "tristate" in its Kconfig and selecting "m"
+for modular builds, however, for modules, exclusively built as LLEXT objects,
+this might not be necessary. Thus it is possible to implement third party Module
```

Comment:
By "not necessary" you mean "not possible"? As in "exclusively". I think I'm guessing what you mean but this is confusing.

Reply:
@marc-hb no, I didn't mean that. I don't know if anybody would want to write modules that they never need to build into the actual SOF image.

Reply:
I'm still confused sorry. I guess I don't understand what is a "module exclusively built as LLEXT object". Sounds like you select "monolithic" but that is ignored by some modules? Why?

Reply:
@marc-hb no, that isn't what I meant. I just meant, that if you're sure to only build your code as a loadable LLEXT module, it doesn't matter what you do in your Kconfig, you can just hard fix it to always build as a module.

Reply:
I still have no idea how you "hard fix" the build-as-module and how the Kconfig gets ignored in that case (which is confusing?)

Reply:
@marc-hb this is what I currently use to switch to building a loadable module: https://github.com/thesofproject/sof/blob/fdb67ee782808f6d5488a53441cdecd75f458677/zephyr/CMakeLists.txt#L874 If you only ever build your code as such, you don't need that `if`, you use that code unconditionally.